dartdoc_test/lib/src/reporter.dart

```diff
@@ -0,0 +1,76 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'package:source_span/source_span.dart';
+import 'package:test/test.dart';
+
+abstract class Reporter {
+  void reportSourceFile(String filename, void Function(SourceFileReporter) fn);
+}
+
+abstract class SourceFileReporter {
+  void reportCodeSample(
+      String identifier, int number, void Function(CodeSampleReporter) fn);
+}
+
+abstract class CodeSampleReporter {
+  void issues(SourceSpan span, String message);
+}
+
+final class ReporterForStdout
```

Comment:
not important, this just my personal preference for structure.

I would do:
```
final class _ReporterForStdout implements ... { ... }
final class _RepoterForTest implements ... { ... }
```
So that they are private, then I would make static methods on `Reporter`, like:
```
abstract class Reporter {
   ...

   /// Create a reporter that prints to stdout
   static Reporter stdout() => _ReporterForStdout();

   /// Create a reporter that registers tests in `package:test`
   static Reporter test() => _RepoterForTest();
}
```

Again, this is just my personal preference. I prefer this because when someone goes to read the generated documentation (created by dartdoc), they won't see all these classes that they don't need to know about. They don't need to know about `RepoterForTest`.

If someone was to write a function that takes a `Reporter` as argument, they should accept `Reporter`, if they accidentally the `ReporterForStdout` instead, then that may work if they only need to output for stdout. But it'll stop working for other scenarios.

If someone writes a function that takes a reporter as argument, example: `void myFunction(Reporter r) {...` that's great. But I don't want them to be able to do `if (r is RepoterForTest) {...` and special-case their test dependent on the kind of `Reporter` they are given.

I'm guessing that nobody will actually write such functions, but we never know. Mostly, when making APIs I strive to make sure that users only have access to the things they need, and that everything they can see (or access) as a meaning.

Again, this is just like my opinion, it's hard to say that there is a right and a wrong here. There are many ways to do things.